pkg/daemon/ceph/osd/volume.go

```diff
@@ -0,0 +1,202 @@
+	storeFlag := "--bluestore"
+	if a.storeConfig.StoreType == config.Filestore {
+		storeFlag = "--filestore"
+	}
+
+	baseArgs := []string{"lvm", "batch", "--prepare", storeFlag, "--yes"}
+	if a.storeConfig.EncryptedDevice {
+		baseArgs = append(baseArgs, encryptedFlag)
+	}
+
+	batchArgs := append(baseArgs, []string{
+		osdsPerDeviceFlag,
+		strconv.Itoa(a.storeConfig.OSDsPerDevice),
+	}...)
+
+	// ceph-volume is soon implementing a parameter to specify the "fast devices", which correspond to the "metadataDevice" from the
+	// crd spec. After that is implemented, we can implement this. In the meantime, we fall back to use rook's partitioning.
+	metadataDeviceSpecified := false
+
+	configured := 0
+	for name, device := range devices.Entries {
+		if device.LegacyPartitionsFound {
+			logger.Infof("skipping device %s configured with legacy rook osd", name)
+			continue
+		}
+
+		if device.Data == -1 {
+			logger.Infof("configuring new device %s", name)
+			deviceArg := path.Join("/dev", name)
+			if metadataDeviceSpecified {
+				// the device will be configured as a batch at the end of the method
+				batchArgs = append(batchArgs, deviceArg)
+				configured++
+			} else {
+				// execute ceph-volume immediately with the device-specific setting instead of batching up multiple devices together
+				immediateExecuteArgs := append(baseArgs, []string{
+					deviceArg,
+					osdsPerDeviceFlag,
+					strconv.Itoa(device.Config.OSDsPerDevice),
+				}...)
+
+				if err := context.Executor.ExecuteCommand(false, "", cephVolumeCmd, immediateExecuteArgs...); err != nil {
+					return fmt.Errorf("failed ceph-volume. %+v", err)
+				}
+
+			}
+		} else {
+			logger.Infof("skipping device %s with osd %d already configured", name, device.Data)
+		}
+	}
+
+	if configured > 0 {
+		if err := context.Executor.ExecuteCommand(false, "", cephVolumeCmd, batchArgs...); err != nil {
+			return fmt.Errorf("failed ceph-volume. %+v", err)
+		}
+	}
+
+	return nil
+}
+func getCephVolumeSupported(context *clusterd.Context) (bool, error) {
+	_, err := context.Executor.ExecuteCommandWithOutput(false, "", cephVolumeCmd, "lvm", "batch", "--prepare")
+	if err != nil {
+		if cmdErr, ok := err.(*exec.CommandError); ok {
+			exitStatus := cmdErr.ExitStatus()
+			if exitStatus == int(syscall.ENOENT) || exitStatus == int(syscall.EPERM) {
+				logger.Infof("supported version of ceph-volume not available")
+				return false, nil
+			}
+			return false, fmt.Errorf("unknown return code from ceph-volume when checking for compatibility: %d", exitStatus)
+		}
+		return false, fmt.Errorf("unknown ceph-volume failure. %+v", err)
+	}
+
+	return true, nil
+}
+
+func getCephVolumeOSDs(context *clusterd.Context, clusterName string) ([]oposd.OSDInfo, error) {
+	result, err := context.Executor.ExecuteCommandWithOutput(false, "", cephVolumeCmd, "lvm", "list", "--format", "json")
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve ceph-volume results. %+v", err)
+	}
+	logger.Debug(result)
```

Comment:
Wouldn't it be better to debug print the result even on error? (e.g., moving it before the `if err != nil` check)